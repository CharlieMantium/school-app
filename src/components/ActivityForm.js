import React from 'react';

export default class ActivityForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      classNo: props.activity ? props.activity.classNo : '',
      day: props.activity ? props.activity.day : '',
      name: props.activity ? props.activity.name : '',
      room: props.activity ? props.activity.room : '',
      teacher: props.activity ? props.activity.teacher : '',
      error: ''
    };
  };
  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => ({name}));
  };
  onTeacherChange = (e) => {
    const teacher = e.target.value;
    this.setState(() => ({teacher}));
  };
  onRoomChange = (e) => {
    const room = e.target.value;
    this.setState(() => ({room}));
  };
  onClassNoChange = (e) => {
    const classNo = e.target.value;
    if (!classNo || classNo.match(/^[1-9]\d*$/)) {
      this.setState(() => ({classNo}));
    };
  };
  onDayChange = (e) => {
    const day = e.target.value;
    this.setState(() => ({day}));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.name || !this.state.day || !this.state.classNo) {
      this.setState(() => ({error: 'Please provide Activity Name, Day and Class Number!'}))
    } else {
      this.setState(() => ({error: ''}))
      this.props.onSubmit({
        classNo: this.state.classNo,
        day: this.state.day,
        name: this.state.name,
        room: this.state.room,
        teacher: this.state.teacher,
      })
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Activity Name"
            autoFocus
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <input
            type="text"
            placeholder="Teacher"
            value={this.state.teacher}
            onChange={this.onTeacherChange}
          />
          <input
            type="text"
            placeholder="Room"
            value={this.state.room}
            onChange={this.onRoomChange}
          />
          <input
            type="number"
            placeholder="Class Number"
            value={this.state.classNo}
            onChange={this.onClassNoChange}
          />
          <select
            value={this.state.day}
            onChange={this.onDayChange}
            required={true}
          >
            <option value=''>Pick a day</option>
            <option value={1}>Monday</option>
            <option value={2}>Tuesday</option>
            <option value={3}>Wednesday</option>
            <option value={4}>Thursday</option>
            <option value={5}>Friday</option>
            <option value={6}>Saturday</option>
            <option value={7}>Sunday</option>
          </select>
          <button>Add Activity</button>
        </form>
      </div>
    )
  }
}
