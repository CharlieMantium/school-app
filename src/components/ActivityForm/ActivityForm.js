import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import styled from 'styled-components';

import { validatePositive, validateDot } from 'helpers/validators';
import { Button } from 'styles/elements/Button';
import { daysOfTheWeek } from 'constants/dates';
import { spacing } from 'styles/base';
import Input from '../Input';

const SelectAndButtonWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  padding: ${spacing.xlSize} 0;
`;

export default class ActivityForm extends React.Component {
  static propTypes = {
    activity: PropTypes.shape({
      activityOrdinal: PropTypes.string,
      day: PropTypes.string,
      id: PropTypes.string,
      name: PropTypes.string,
      room: PropTypes.string,
      teacher: PropTypes.string,
    }),
    onSubmit: PropTypes.func.isRequired,
  };

  static defaultProps = {
    activity: {
      activityOrdinal: '',
      day: '',
      name: '',
      room: '',
      teacher: '',
    },
  };

  constructor(props) {
    super(props);
    this.state = {
      activityOrdinal: props.activity.activityOrdinal,
      day: props.activity.day,
      name: props.activity.name,
      room: props.activity.room,
      teacher: props.activity.teacher,
      activityOrdinalError: '',
      nameError: '',
      roomError: '',
      teacherError: '',
    };
  }

  onInputValueChange = field => e => {
    const inputValue = e.target.value;
    if (field === 'activityOrdinal') {
      if (validatePositive(inputValue) && validateDot(inputValue)) {
        this.setState({
          [field]: inputValue,
          [`${field}Error`]: '',
        });
      }
    } else {
      this.setState({
        [field]: inputValue,
        [`${field}Error`]: '',
      });
    }
  };

  onBlur = field => e => !e.target.value && this.setState({ [field]: 'Required!' });

  onSubmit = e => {
    e.preventDefault();
    this.setState({
      nameError: '',
      activityOrdinalError: '',
      roomError: '',
      teacherError: '',
    });
    const { name, day, activityOrdinal, room, teacher } = this.state;
    const errors = {};
    const errorMsg = 'Required!';
    if (!name || !day || !activityOrdinal || !teacher || !room) {
      if (!name) {
        errors.nameError = errorMsg;
      }
      if (!activityOrdinal) {
        errors.activityOrdinalError = errorMsg;
      }
      if (!room) {
        errors.roomError = errorMsg;
      }
      if (!teacher) {
        errors.teacherError = errorMsg;
      }
      if (!_.isEmpty(errors)) {
        this.setState(errors);
      }
    } else {
      this.props.onSubmit({
        activityOrdinal,
        day,
        name,
        room,
        teacher,
      });
    }
  };

  render() {
    const {
      name,
      day,
      activityOrdinal,
      room,
      teacher,
      activityOrdinalError,
      nameError,
      roomError,
      teacherError,
    } = this.state;
    return (
      <form onSubmit={this.onSubmit} data-test="form">
        <Input
          type="text"
          placeholder="Activity Name"
          value={name}
          onChange={this.onInputValueChange('name')}
          onBlur={this.onBlur('nameError')}
          errorMsg={nameError}
          data-test="input-component-name"
        />
        <Input
          type="text"
          placeholder="Teacher"
          value={teacher}
          onChange={this.onInputValueChange('teacher')}
          onBlur={this.onBlur('teacherError')}
          errorMsg={teacherError}
          data-test="input-component-teacher"
        />
        <Input
          type="text"
          placeholder="Room"
          value={room}
          onChange={this.onInputValueChange('room')}
          onBlur={this.onBlur('roomError')}
          errorMsg={roomError}
          data-test="input-component-room"
        />
        <Input
          type="number"
          step="1"
          placeholder="Class Number"
          value={activityOrdinal}
          onChange={this.onInputValueChange('activityOrdinal')}
          onBlur={this.onBlur('activityOrdinalError')}
          errorMsg={activityOrdinalError}
          data-test="input-component-activityOrdinal"
        />
        <SelectAndButtonWrapper>
          <select
            value={day}
            onChange={this.onInputValueChange('day')}
            required
            data-test="select-day"
          >
            <option value="" data-test="select-option-default">
              Pick a day
            </option>
            {daysOfTheWeek.map(weekDay => (
              <option value={weekDay} key={weekDay} data-test={`select-option-${weekDay}`}>
                {weekDay}
              </option>
            ))}
          </select>
          {/* TODO: You should start to think about separate Dropdown component. It's a hard thing to do,
          using native html, it's harder to style options, etc. You could build it by yourself,
          but also you can use on of the most popular react packages (atm) - rc-dropdown or downshift.
          It will be big exercise. So, do it later (when you start real styling - do it also quite asap).
          But by separate pr only for dropdown. */}
          <Button type="submit" data-test="button-submit">
            {this.props.activity.name ? 'Update Activity' : 'Add Activity'}
          </Button>
        </SelectAndButtonWrapper>
      </form>
    );
  }
}
