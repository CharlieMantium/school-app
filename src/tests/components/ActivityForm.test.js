import React from 'react';
import { shallow } from 'enzyme';

import ActivityForm from '../../components/ActivityForm';
import testState from '../fixtures/state';

test('should render ActivityForm correctly', () => {
  const wrapper = shallow(<ActivityForm onSubmit={() => {}} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ActivityForm with testState', () => {
  const wrapper = shallow(
    <ActivityForm on onSubmit={() => {}} activity={testState.activities.items[1]} />,
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submition', () => {
  const wrapper = shallow(<ActivityForm onSubmit={() => {}} />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {},
  });
  expect(wrapper.state('classNoError').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});
