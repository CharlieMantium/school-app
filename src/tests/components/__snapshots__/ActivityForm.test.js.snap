// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ActivityForm correctly 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Activity Name"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Teacher"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Room"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Class Number"
      step="1"
      type="number"
      value=""
    />
    <select
      className="form__select"
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        className="form__selectOption"
        value=""
      >
        Pick a day
      </option>
      <option
        className="form__selectOption"
        key="monday"
        value="monday"
      >
        monday
      </option>
      <option
        className="form__selectOption"
        key="tuesday"
        value="tuesday"
      >
        tuesday
      </option>
      <option
        className="form__selectOption"
        key="wednesday"
        value="wednesday"
      >
        wednesday
      </option>
      <option
        className="form__selectOption"
        key="thursday"
        value="thursday"
      >
        thursday
      </option>
      <option
        className="form__selectOption"
        key="friday"
        value="friday"
      >
        friday
      </option>
      <option
        className="form__selectOption"
        key="saturday"
        value="saturday"
      >
        saturday
      </option>
      <option
        className="form__selectOption"
        key="sunday"
        value="sunday"
      >
        sunday
      </option>
    </select>
    <button
      type="submit"
    >
      Add Activity
    </button>
  </form>
</div>
`;

exports[`should render ActivityForm with testState 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Activity Name"
      type="text"
      value="polski"
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Teacher"
      type="text"
      value="Malski"
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Room"
      type="text"
      value="hell"
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Class Number"
      step="1"
      type="number"
      value="1"
    />
    <select
      className="form__select"
      onChange={[Function]}
      required={true}
      value="monday"
    >
      <option
        className="form__selectOption"
        value=""
      >
        Pick a day
      </option>
      <option
        className="form__selectOption"
        key="monday"
        value="monday"
      >
        monday
      </option>
      <option
        className="form__selectOption"
        key="tuesday"
        value="tuesday"
      >
        tuesday
      </option>
      <option
        className="form__selectOption"
        key="wednesday"
        value="wednesday"
      >
        wednesday
      </option>
      <option
        className="form__selectOption"
        key="thursday"
        value="thursday"
      >
        thursday
      </option>
      <option
        className="form__selectOption"
        key="friday"
        value="friday"
      >
        friday
      </option>
      <option
        className="form__selectOption"
        key="saturday"
        value="saturday"
      >
        saturday
      </option>
      <option
        className="form__selectOption"
        key="sunday"
        value="sunday"
      >
        sunday
      </option>
    </select>
    <button
      type="submit"
    >
      Add Activity
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submition 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Activity Name"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Teacher"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Room"
      type="text"
      value=""
    />
    <Input
      errorMsg=""
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Class Number"
      step="1"
      type="number"
      value=""
    />
    <select
      className="form__select"
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        className="form__selectOption"
        value=""
      >
        Pick a day
      </option>
      <option
        className="form__selectOption"
        key="monday"
        value="monday"
      >
        monday
      </option>
      <option
        className="form__selectOption"
        key="tuesday"
        value="tuesday"
      >
        tuesday
      </option>
      <option
        className="form__selectOption"
        key="wednesday"
        value="wednesday"
      >
        wednesday
      </option>
      <option
        className="form__selectOption"
        key="thursday"
        value="thursday"
      >
        thursday
      </option>
      <option
        className="form__selectOption"
        key="friday"
        value="friday"
      >
        friday
      </option>
      <option
        className="form__selectOption"
        key="saturday"
        value="saturday"
      >
        saturday
      </option>
      <option
        className="form__selectOption"
        key="sunday"
        value="sunday"
      >
        sunday
      </option>
    </select>
    <button
      type="submit"
    >
      Add Activity
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submition 2`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <Input
      errorMsg="Required"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Activity Name"
      type="text"
      value=""
    />
    <Input
      errorMsg="Required"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Teacher"
      type="text"
      value=""
    />
    <Input
      errorMsg="Required"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Room"
      type="text"
      value=""
    />
    <Input
      errorMsg="Required"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Class Number"
      step="1"
      type="number"
      value=""
    />
    <select
      className="form__select"
      onChange={[Function]}
      required={true}
      value=""
    >
      <option
        className="form__selectOption"
        value=""
      >
        Pick a day
      </option>
      <option
        className="form__selectOption"
        key="monday"
        value="monday"
      >
        monday
      </option>
      <option
        className="form__selectOption"
        key="tuesday"
        value="tuesday"
      >
        tuesday
      </option>
      <option
        className="form__selectOption"
        key="wednesday"
        value="wednesday"
      >
        wednesday
      </option>
      <option
        className="form__selectOption"
        key="thursday"
        value="thursday"
      >
        thursday
      </option>
      <option
        className="form__selectOption"
        key="friday"
        value="friday"
      >
        friday
      </option>
      <option
        className="form__selectOption"
        key="saturday"
        value="saturday"
      >
        saturday
      </option>
      <option
        className="form__selectOption"
        key="sunday"
        value="sunday"
      >
        sunday
      </option>
    </select>
    <button
      type="submit"
    >
      Add Activity
    </button>
  </form>
</div>
`;
